### ISSUE TITLE
Error on empty message-data

### ISSUE THREAD

**Issue by jacob418** [kreait/firebase-php#591]
When providing an ampty array for `CloudMessage::withData` the message send to the Google-API will contain `{ message: { data: [] } }` - this is not acceptable according tho the error-message from the debug log: 

<details>
<summary>
Debug-log
</summary>

```
POST /batch HTTP/1.1
User-Agent: GuzzleHttp/6.5.5 curl/7.70.0 PHP/7.4.8
Host: fcm.googleapis.com
Content-Type: multipart/mixed; boundary=4098b7e719e70ab2565b58eb39617ca791f426ed
Content-Length: 2080

--4098b7e719e70ab2565b58eb39617ca791f426ed
Content-Length: 813
Content-ID: 1
Content-Transfer-Encoding: binary
Content-Type: application/http

POST /v1/projects/xxx/messages:send HTTP/1.1

{"message":{"data":[],"notification":{"title":"New Message from Mustafa Hartmann","body":"My Message!"},"android":{"ttl":"3600s","priority":"normal","notification":{"title":"New Message from Mustafa Hartmann","body":"My Message!","icon":"stock_ticker_update","color":"#f45342"}},"apns":{"headers":{"apns-priority":"10"},"payload":{"aps":{"alert":{"title":"New Message from Mustafa Hartmann","body":"My Message!"},"badge":42}}},"webpush":{"notification":{"title":"New Message from Mustafa Hartmann","body":"My Message!","icon":"https:\/\/my-server\/icon.png"},"fcm_options":{"link":"https:\/\/my-server\/some-page"}},"token":"xxx"},"validate_only":false}
--4098b7e719e70ab2565b58eb39617ca791f426ed
Content-Length: 813
Content-ID: 2
Content-Transfer-Encoding: binary
Content-Type: application/http

POST /v1/projects/app1start/messages:send HTTP/1.1

{"message":{"data":[],"notification":{"title":"New Message from Mustafa Hartmann","body":"My Message!"},"android":{"ttl":"3600s","priority":"normal","notification":{"title":"New Message from Mustafa Hartmann","body":"My Message!","icon":"stock_ticker_update","color":"#f45342"}},"apns":{"headers":{"apns-priority":"10"},"payload":{"aps":{"alert":{"title":"New Message from Mustafa Hartmann","body":"My Message!"},"badge":42}}},"webpush":{"notification":{"title":"New Message from Mustafa Hartmann","body":"My Message!","icon":"https:\/\/my-server\/icon.png"},"fcm_options":{"link":"https:\/\/my-server\/some-page"}},"token":"xxx"},"validate_only":false}
--4098b7e719e70ab2565b58eb39617ca791f426ed--
<<<<<<<<
HTTP/1.1 200 OK
Content-Type: multipart/mixed; boundary=batch_xm1DaJI91r7KcR1e9sZO0jZo0gk_qFAF
Vary: X-Origin, Referer, Origin,Accept-Encoding
Date: Wed, 21 Apr 2021 12:45:11 GMT
Server: ESF
Cache-Control: private
X-XSS-Protection: 0
X-Frame-Options: SAMEORIGIN
X-Content-Type-Options: nosniff
Alt-Svc: h3-29=":443"; ma=2592000,h3-T051=":443"; ma=2592000,h3-Q050=":443"; ma=2592000,h3-Q046=":443"; ma=2592000,h3-Q043=":443"; ma=2592000,quic=":443"; ma=2592000; v="46,43"
Accept-Ranges: none
Transfer-Encoding: chunked


--batch_xm1DaJI91r7KcR1e9sZO0jZo0gk_qFAF
Content-Type: application/http
Content-ID: response-1

HTTP/1.1 400 Bad Request
Vary: Origin
Vary: X-Origin
Vary: Referer
Content-Type: application/json; charset=UTF-8

{
  "error": {
    "code": 400,
    "message": "Invalid value at 'message' (Map), Cannot bind a list to map for field 'data'.",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.BadRequest",
        "fieldViolations": [
          {
            "field": "message",
            "description": "Invalid value at 'message' (Map), Cannot bind a list to map for field 'data'."
          }
        ]
      }
    ]
  }
}

--batch_xm1DaJI91r7KcR1e9sZO0jZo0gk_qFAF
Content-Type: application/http
Content-ID: response-2

HTTP/1.1 400 Bad Request
Vary: Origin
Vary: X-Origin
Vary: Referer
Content-Type: application/json; charset=UTF-8

{
  "error": {
    "code": 400,
    "message": "Invalid value at 'message' (Map), Cannot bind a list to map for field 'data'.",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.BadRequest",
        "fieldViolations": [
          {
            "field": "message",
            "description": "Invalid value at 'message' (Map), Cannot bind a list to map for field 'data'."
          }
        ]
      }
    ]
  }
}

--batch_xm1DaJI91r7KcR1e9sZO0jZo0gk_qFAF--

--------
NULL  
```
</details>

Providing at least one key in the array passed to `CloudMessage::withData` will solve the problem. As far as I experienced `Array<->stdClass` conversations in PHP this is a known behavior. The workaround would be a cast to `object` in `MessageData::jsonSerialize:48` and return `stdClass`:

```php
    /**
     * @return stdClass
     */
    public function jsonSerialize(): stdClass
    {
        return (object) $this->data;
    }
```

Also it would be nice to be able to pass `stdClass` directly for `CloudMessage::withData` - however this is prohibited by the PHP type system (`MessageData|array<string, string>` does not allow `stdClass`/JSON).

---

Working on windows 10x64

<details>
<summary>
PHP/Composer version

```
PS C:\Users\xxx> php -v
PHP 7.4.8 (cli) (built: Jul  9 2020 11:30:39) ( ZTS Visual C++ 2017 x64 )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Xdebug v2.9.6, Copyright (c) 2002-2020, by Derick Rethans
PS C:\Users\xxx> composer -V
Composer version 2.0.9 2021-01-27 16:09:27
```
</details>

<details>
<summary>
composer.json:
</summary>

```
{
    "name": "laravel/laravel",
    "type": "project",
    "version": "0.1.0",
    "description": "The Laravel Framework.",
    "keywords": [
        "framework",
        "laravel"
    ],
    "license": "MIT",
    "require": {
        "php": "^7.4.0",
        "chriskonnertz/open-graph": "^2.0",
        "fideloper/proxy": "^4.2",
        "fruitcake/laravel-cors": "^2.0",
        "guzzlehttp/guzzle": "^6.3",
        "kreait/laravel-firebase": "^3.1",
        "laravel/framework": "^7.0",
        "laravel/sanctum": "^2.4",
        "laravel/tinker": "^2.0",
        "musonza/chat": "^4.4",
        "spatie/geocoder": "^3.10",
        "spatie/laravel-medialibrary": "^8.3.0"
    },
    "require-dev": {
        "barryvdh/laravel-ide-helper": "^2.7",
        "facade/ignition": "^2.0",
        "fzaninotto/faker": "^1.9.1",
        "mockery/mockery": "^1.3.1",
        "nunomaduro/collision": "^4.1",
        "phpunit/phpunit": "^8.5",
        "squizlabs/php_codesniffer": "^3.5"
    },
    "config": {
        "optimize-autoloader": true,
        "preferred-install": "dist",
        "sort-packages": true
    },
    "extra": {
        "laravel": {
            "dont-discover": []
        }
    },
    "autoload": {
        "psr-4": {
            "App\\": "app/"
        },
        "classmap": [
            "database/seeds",
            "database/factories"
        ]
    },
    "autoload-dev": {
        "psr-4": {
            "Tests\\": "tests/"
        }
    },
    "minimum-stability": "dev",
    "prefer-stable": true,
    "scripts": {
        "post-autoload-dump": [
            "Illuminate\\Foundation\\ComposerScripts::postAutoloadDump",
            "@php artisan package:discover --ansi",
            "@php artisan ide-helper:generate",
            "@php artisan ide-helper:meta"
        ],
        "post-root-package-install": [
            "@php -r \"file_exists('.env') || copy('.env.example', '.env');\""
        ],
        "post-create-project-cmd": [
            "@php artisan key:generate --ansi"
        ],
        "ide-helper": [
            "@php artisan ide-helper:generate",
            "@php artisan ide-helper:meta",
            "php artisan ide-helper:models -N"
        ],
        "post-update-cmd": [
            "@php artisan cache:clear"
        ],
        "post-install-cmd": [
            "@php artisan cache:clear"
        ]
    }
}
```
</details>

**Comment by jeromegamez** [comment]
I was able to confirm this with a failing test, thanks for bringing this up and your input 🙏 .

Casting to an object instead of keeping the array would break backward compatibility, so I'll probably go with a slightly more complicated than necessary check when serializing the message envelope. This has the benefit that the `data` field doesn't need to be included at all if it's empty anyway.

You can track progress on this at #592. 

PS: Why did you give the message empty data 😅 
PPS: I'm afraid I will not add support for `stdClass`/`object` as method arguments, they're somewhat harder to check than arrays (e.g. you could then pass a `DateTime` as a "valid" parameter and what would I be supposed to do with that 😅. You can always `(array) $myObject` or `json_decode($myJson)`, but I suppose that's what you're already doing ^^

**Comment by jacob418** [comment]
> PS: Why did you give the message empty data 😅

It was first iteration of implementation. I just added some methods to my base Notification model to collect data and then format it. This way I have to write code for generating fcm only once. Since it was the first iteration I didnt think much about data at this point and just passed an empty array as placeholder. But of course the final version will skip data if its empty.

> PPS: I'm afraid I will not add support for stdClass/object as method arguments, they're somewhat harder to check than arrays (e.g. you could then pass a DateTime as a "valid" parameter and what would I be supposed to do with that 😅. You can always (array) $myObject or json_decode($myJson), but I suppose that's what you're already doing ^^

Thats ok, it was just an asumption what might be nice to have. Btw, does `array<string, string>` allow numbers/boolean/null as value? I guess not?!

**Comment by jeromegamez** [comment]
I saw your issue on the laravel package, just wasn't able to look into it yet.

Your assumption about MessageData only accepting `array<string, string>` is correct. Firebase doesn't accept non-strings in the message data, and if the MessageData class would allow them, it would have to cast them to perhaps unexpected results, leading to questions like "I defined it to be true, why does it arrive on the device as `'true'` or `'1'`" - better be explicit about what to expect to begin with 😅

**Comment by jacob418** [comment]
> better be explicit about what to expect to begin with 😅

I totally agree, I wasnt aware of that limitation.
